crates/miden-objects/src/account/component/template/storage/placeholder.rs

```diff
@@ -148,13 +149,63 @@ pub enum StorageValueNameError {
 #[derive(Debug, Error)]
 pub enum TemplateTypeError {
     #[error("failed to parse input `{0}` as `{1}`")]
-    ParseError(String, String),
+    ParseError(String, TemplateType),
     #[error("conversion error: {0}")]
     ConversionError(String),
     #[error("felt type ` {0}` not found in the type registry")]
-    FeltTypeNotFound(String),
+    FeltTypeNotFound(TemplateType),
     #[error("word type ` {0}` not found in the type registry")]
-    WordTypeNotFound(String),
+    WordTypeNotFound(TemplateType),
+}
+
+// TEMPLATE TYPE
+// ================================================================================================
+
+/// A newtype wrapper around a `String`, representing a template's type identifier.
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+#[cfg_attr(feature = "std", derive(::serde::Deserialize, ::serde::Serialize))]
+#[cfg_attr(feature = "std", serde(transparent))]
+pub struct TemplateType(String);
+
+impl TemplateType {
+    /// Creates a new [`TemplateType`] from a `String``.
+    pub fn new(s: impl Into<String>) -> Self {
+        Self(s.into())
+    }
+
+    /// Returns the default felt type,  assigned to entries with no type.
+    pub fn default_felt_type() -> TemplateType {
+        TemplateType::new("felt")
+    }
+
+    /// Returns the default felt type, assigned to entries with no type.
+    pub fn default_word_type() -> TemplateType {
+        TemplateType::new("word")
+    }
+
+    /// Returns a reference to the inner string.
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl Display for TemplateType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl Serializable for TemplateType {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write(self.0.clone())
+    }
+}
+
+impl Deserializable for TemplateType {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let id: String = source.read()?;
+        Ok(TemplateType::new(id))
+    }
 }
```

Comment:
Unfortunately could not make `TemplateType` contain an inner `'static &str` due to serialization, unless we wanted deserialization to leak memory

Reply:
I guess it would be possible with an inner representation of `Cow<'static, str>`? But not essential I think.